File: app/src/main/java/com/alexkn/syntact/app/ApplicationComponent.java
package com.alexkn.syntact.app;

import android.content.Context;

import com.alexkn.syntact.domain.usecase.bucket.CreateBucketWorker;
import com.alexkn.syntact.domain.usecase.play.FetchPhrasesWorker;
import com.alexkn.syntact.presentation.bucketlist.PlayMenuViewModel;
import com.alexkn.syntact.presentation.createbucket.CreateBucketViewModel;
import com.alexkn.syntact.presentation.flashcard.FlashcardViewModel;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;

@Component(modules = {ApplicationModule.class})
@Singleton
public interface ApplicationComponent {

    @Component.Builder
    interface Builder {

        @BindsInstance
        Builder applicationContext(Context applicationContext);

        ApplicationComponent build();
    }

    void inject(FetchPhrasesWorker fetchPhrasesWorker);

    void inject(CreateBucketWorker createBucketWorker);

    ViewModelFactory<PlayMenuViewModel> playMenuViewModelFactory();

    ViewModelFactory<CreateBucketViewModel> createBucketViewModelFactory();

    ViewModelFactory<FlashcardViewModel> flashcardViewModelFactory();
}

File: app/src/main/java/com/alexkn/syntact/app/ApplicationComponent.kt
package com.alexkn.syntact.app

import android.content.Context
import com.alexkn.syntact.data.config.DaoModule
import com.alexkn.syntact.data.config.DatabaseModule
import com.alexkn.syntact.presentation.common.ViewModelFactory
import com.alexkn.syntact.presentation.deckboard.DeckBoardViewModel
import com.alexkn.syntact.presentation.deckcreation.DeckCreationViewModel
import com.alexkn.syntact.presentation.deckdetails.DeckDetailsViewModel
import com.alexkn.syntact.presentation.decklist.DeckListViewModel
import dagger.BindsInstance
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(modules = [NetworkModule::class, DatabaseModule::class, DaoModule::class])
interface ApplicationComponent {

    @Component.Builder
    interface Builder {

        @BindsInstance
        fun applicationContext(applicationContext: Context): Builder

        fun build(): ApplicationComponent
    }

    fun playMenuViewModelFactory(): ViewModelFactory<DeckListViewModel>

    fun flashcardViewModelFactory(): ViewModelFactory<DeckBoardViewModel>

    fun bucketDetailsViewModelFactory(): ViewModelFactory<DeckDetailsViewModel>

    fun createTemplateViewModelFactory(): ViewModelFactory<DeckCreationViewModel>
}

interface ApplicationComponentProvider {
    val applicationComponent: ApplicationComponent
}